[WTF] Implement convenience functions for CFArray <-> Vector conversions

ddkilzer opened the pull request:
#### 2f9450c601c6090870dcfe81c870bc54d880ebda
<pre>
[WTF] Implement convenience functions for CFArray &lt;-&gt; Vector conversions
<a href="https://bugs.webkit.org/show_bug.cgi?id=245284">https://bugs.webkit.org/show_bug.cgi?id=245284</a>
&lt;rdar://100032945&gt;

Reviewed by Darin Adler.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformMac.cmake:
* Source/WTF/wtf/PlatformWin.cmake:
- Add VectorCF.h to project files.

* Source/WTF/wtf/cf/VectorCF.h: Copied from Source/WTF/wtf/cocoa/VectorCocoa.h.
- Add convenience functions for CFArray &lt;-&gt; Vector conversions.
(WTF::ParameterTypeTraits): Add.
- Get the type of the first parameter of a lambda.
(WTF::LambdaTypeTraits): Add.
- Get type traits for a lambda like the return type and
  parameter count.
(WTF::addUnlessNil): Add.
(WTF::createCFArray): Add.
- This uses Checked&lt;&gt; to convert from size_t to CFIndex.
(WTF::makeVector&lt;String&gt;): Add specialization.
(WTF::makeVector): Add.
- Use dynamic_cf_cast&lt;&gt;() to avoid having to do it at the call
  site, and take a CFType template parameter so that helper
  functions use a specific CF type.
(WTF::vectorFromCFData): Add.
- Use Checked&lt;&gt; to convert from CFData to Vector&lt;uint8_t&gt;.
(WTF::makeVectorElement): Add.
(WTF::makeCFArrayElement): Add.
- Add convenience function for float &lt;-&gt; CFNumberRef.

* Source/WTF/wtf/text/WTFString.h:
(WTF::makeCFArrayElement):
(WTF::makeVectorElement):
* Source/WTF/wtf/text/cf/StringCF.cpp:
(WTF::makeCFArrayElement):
(WTF::makeVectorElement):
- Add helper functions for CFString &lt;-&gt; String conversions.

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::platformPreferredLanguages):
(WebCore::CaptionUserPreferencesMediaAF::preferredAudioCharacteristics const):
- Use makeVector&lt;String&gt;().
* Source/WebCore/platform/graphics/ca/win/PlatformCAAnimationWin.cpp:
(PlatformCAAnimationWin::setValues):
(PlatformCAAnimationWin::setKeyTimes):
(PlatformCAAnimationWin::setTimingFunctions):
(PlatformCAAnimationWin::setAnimations):
- Use createCFArray().
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCResolverCocoa.cpp:
(WebKit::resolvedName):
- Use makeVector&lt;WebCore::IPAddress, CFDataRef&gt;().
* Source/WebKit/Shared/cf/CookieStorageUtilsCF.mm:
(WebKit::identifyingDataFromCookieStorage):
- Use vectorFromCFData().
* Source/WebKitLegacy/win/WebHistoryItem.cpp:
(WebHistoryItem::initFromDictionaryRepresentation):
- Use makeVector&lt;String&gt;().

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
- Add missing RetainPtrHashing.cpp tests!
- Move tests for code in WTF into the TestWTF target.
- Add VectorCF.cpp tests.
* Tools/TestWebKitAPI/Tests/WTF/cf/VectorCF.cpp: Add.
(TestWebKitAPI::TEST):
- Add tests for VectorCF.h.

Canonical link: <a href="https://commits.webkit.org/254665@main">https://commits.webkit.org/254665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf6d9092f0d939a1a64ad7ec9e0b569d99aed0f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89740 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34291 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20452 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99085 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155901 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93748 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32785 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28248 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82109 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93435 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95387 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26054 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76585 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26007 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80931 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80796 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69001 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81402 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30537 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14880 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76261 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30286 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15817 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26431 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3279 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38763 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/78847 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32447 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34905 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17278 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/ca/win/PlatformCAAnimationWin.cpp:
```diff
@@ -443,63 +444,48 @@ void PlatformCAAnimationWin::copyToValueFrom(const PlatformCAAnimation& value)
-    CACFAnimationSetValues(m_animation.get(), array.get());
+    CACFAnimationSetValues(m_animation.get(), createCFArray(values, [] (float value) {
+        float number = value;
+        return adoptCF(CFNumberCreate(0, kCFNumberFloatType, &number))
```
No need to have a local variable. There’s no problem taking a pointer to an argument.

Also, I think we could consider having the default behavior for a `Vector<float>` create a vector of `CFNumber` in this fashion, instead of doing it here.

ddkilzer replied:
Added a convenience function for `Vector<float>` to `CFNumberRef`.

darinadler commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCResolverCocoa.cpp:
```diff
@@ -29,9 +29,15 @@
 
 namespace WebKit {
 
+static std::optional<WebCore::IPAddress> makeVectorElement(const WebCore::IPAddress* const, CFDataRef data)
```
No need for the `const` after the `*`.

ddkilzer replied:
The compiler said otherwise when I was working on the patch!

I changed this to be an inline lambda, though, since it read better.

darinadler commented on Source/WebCore/platform/graphics/ca/win/PlatformCAAnimationWin.cpp:
```diff
@@ -515,18 +501,15 @@ void PlatformCAAnimationWin::copyValuesFrom(const PlatformCAAnimation& value)
-    CACFAnimationSetKeyTimes(m_animation.get(), array.get());
+    CACFAnimationSetKeyTimes(m_animation.get(), createCFArray(values, [] (float value) {
+        float number = value;
+        return adoptCF(CFNumberCreate(0, kCFNumberFloatType, &number))
```
Same comments as above.

darinadler commented on Source/WebKitLegacy/win/WebHistoryItem.cpp:
```diff
@@ -113,12 +115,8 @@ HRESULT WebHistoryItem::initFromDictionaryRepresentation(_In_opt_ void* dictiona
-            (*redirectURLsVector)[i] = String(static_cast<CFStringRef>(CFArrayGetValueAtIndex(redirectURLsRef, i)));
-    }
+    if (CFArrayRef redirectURLsRef = dynamic_cf_cast<CFArrayRef>(CFDictionaryGetValue(dictionaryRef, redirectURLsKey)))
+        redirectURLsVector = WTFMove(makeVector<String>(redirectURLsRef));
```
This `WTFMove` is not required, return values like this are already rvalue references, and I am surprised it compiles.

ddkilzer replied:
The LHS is actually a `std::unique_ptr<Vector<String>>`, so I'm pretty sure `WTFMove()` was correct here.

Also, WinCairo doesn't build with the `WTFMove()` removed:  <https://ews-build.webkit.org/#/builders/12/builds/93305>

ddkilzer replied:
Actually, I think I needed to use `makeUnique<Vector<String>>(WTFMove(makeVector<String>(redirectURLsRef)))` here.  Trying that now.

ddkilzer replied:
I was wrong.  Removing `WTFMove()`:  <https://ews-build.webkit.org/#/builders/12/builds/93413>

```
C:\BW\WinCairo-EWS\build\WebKitBuild\Release\WTF\Headers\wtf/StdLibExtras.h(528): error C2338: T is not an lvalue reference; move() is unnecessary.
C:\BW\WinCairo-EWS\build\Source\WebKitLegacy\win\WebHistoryItem.cpp(118): note: see reference to function template instantiation 'WTF::Vector<WTF::String,0,WTF::CrashOnOverflow,16,WTF::VectorMalloc> &&std::move<WTF::CheckMoveParameter,WTF::Vector<WTF::String,0,WTF::CrashOnOverflow,16,WTF::VectorMalloc>>(T &&)' being compiled
        with
        [
            T=WTF::Vector<WTF::String,0,WTF::CrashOnOverflow,16,WTF::VectorMalloc>
        ]
```

ddkilzer commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCResolverCocoa.cpp:
```diff
@@ -46,18 +47,12 @@ static void resolvedName(CFHostRef hostRef, CFHostInfoType typeInfo, const CFStr
-    }
+    auto addresses = makeVector<CFDataRef>(resolvedAddresses, [] (CFDataRef data) -> std::optional<WebCore::IPAddress> {
+        return WebCore::IPAddress::fromSockAddrIn6(*reinterpret_cast<const struct sockaddr_in6*>(CFDataGetBytePtr(data)));
+    });
```
I feel like the `makeVector` template for this function should be able to figure out both the `Vector<>` element type and the `CFType` using the functor, but I haven't figured out how to write that in C++ template code.

It's a little non-intuitive (and weird) to write `makeVector<CFDataRef>` here when the output is a `Vector<WebCore::IPAddress>`.  (I could change it back to `makeVector<WebCore::IPAddress, CFDataRef>`, but the functor should have all the type info we need!)

ddkilzer replied:
I think I figured this out.  Had to think differently about how to declare the arguments to `MapFunctionType` in the template.

ddkilzer replied:
Or not.  Trying to figure out how to use a `Function<Out(In...)>` type.

ddkilzer replied:
Figured it out for real using this “one weird trick”.  Patch tomorrow.

ddkilzer replied:
The "trick" I found was to use `decltype(&LambdaType::operator())` to get a parameter pack for a pointer to the lambda's `operator()` method, then use that to get the return type and the list of parameters.  It's brilliant in its simplicity.

From there I used another template (by expanding the parameter pack for the list of parameters) to get the first parameter's type.

This prevents the need to declare the element type of the `Vector<>` or the `CFTypeRef` more than once for this template:
```
template<typename MapLambdaType> Vector<typename LambdaTypeTraits<MapLambdaType>::returnType::value_type> makeVector(CFArrayRef, MapLambdaType&&);
```

When used like this in `Source/WebKit/NetworkProcess/webrtc/NetworkRTCResolverCocoa.cpp`:
```
    auto addresses = makeVector(resolvedAddresses, [] (CFDataRef data) -> std::optional<WebCore::IPAddress> {
        return WebCore::IPAddress::fromSockAddrIn6(*reinterpret_cast<const struct sockaddr_in6*>(CFDataGetBytePtr(data)));
    });
```

The only improvement here would be to add support for passing a `WTF::Function<>` instead, but I think this can be done in a later commit (as there is no need for it in this patch).

Commit:
[WTF] Implement convenience functions for CFArray <-> Vector conversions
https://bugs.webkit.org/show_bug.cgi?id=245284
<rdar://100032945>

Reviewed by Darin Adler.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformMac.cmake:
* Source/WTF/wtf/PlatformWin.cmake:
- Add VectorCF.h to project files.

* Source/WTF/wtf/cf/VectorCF.h: Copied from Source/WTF/wtf/cocoa/VectorCocoa.h.
- Add convenience functions for CFArray <-> Vector conversions.
(WTF::ParameterTypeTraits): Add.
- Get the type of the first parameter of a lambda.
(WTF::LambdaTypeTraits): Add.
- Get type traits for a lambda like the return type and
  parameter count.
(WTF::addUnlessNil): Add.
(WTF::createCFArray): Add.
- This uses Checked<> to convert from size_t to CFIndex.
(WTF::makeVector<String>): Add specialization.
(WTF::makeVector): Add.
- Use dynamic_cf_cast<>() to avoid having to do it at the call
  site, and take a CFType template parameter so that helper
  functions use a specific CF type.
(WTF::vectorFromCFData): Add.
- Use Checked<> to convert from CFData to Vector<uint8_t>.
(WTF::makeVectorElement): Add.
(WTF::makeCFArrayElement): Add.
- Add convenience function for float <-> CFNumberRef.

* Source/WTF/wtf/text/WTFString.h:
(WTF::makeCFArrayElement):
(WTF::makeVectorElement):
* Source/WTF/wtf/text/cf/StringCF.cpp:
(WTF::makeCFArrayElement):
(WTF::makeVectorElement):
- Add helper functions for CFString <-> String conversions.

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::platformPreferredLanguages):
(WebCore::CaptionUserPreferencesMediaAF::preferredAudioCharacteristics const):
- Use makeVector<String>().
* Source/WebCore/platform/graphics/ca/win/PlatformCAAnimationWin.cpp:
(PlatformCAAnimationWin::setValues):
(PlatformCAAnimationWin::setKeyTimes):
(PlatformCAAnimationWin::setTimingFunctions):
(PlatformCAAnimationWin::setAnimations):
- Use createCFArray().
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCResolverCocoa.cpp:
(WebKit::resolvedName):
- Use makeVector<WebCore::IPAddress, CFDataRef>().
* Source/WebKit/Shared/cf/CookieStorageUtilsCF.mm:
(WebKit::identifyingDataFromCookieStorage):
- Use vectorFromCFData().
* Source/WebKitLegacy/win/WebHistoryItem.cpp:
(WebHistoryItem::initFromDictionaryRepresentation):
- Use makeVector<String>().

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
- Add missing RetainPtrHashing.cpp tests!
- Move tests for code in WTF into the TestWTF target.
- Add VectorCF.cpp tests.
* Tools/TestWebKitAPI/Tests/WTF/cf/VectorCF.cpp: Add.
(TestWebKitAPI::TEST):
- Add tests for VectorCF.h.

Canonical link: https://commits.webkit.org/254665@main
